cmd/kube-proxy/app/server_others.go

```diff
@@ -183,29 +181,36 @@ func newProxyServer(
 	iptInterface = utiliptables.New(execer, primaryProtocol)
 
 	var ipt [2]utiliptables.Interface
-	dualStack := utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) && proxyMode != proxyModeUserspace
-	if dualStack {
+	dualStack := true // While we assume that node supports, we do further checks below
+	if proxyMode != proxyModeUserspace {
 		// Create iptables handlers for both families, one is already created
 		// Always ordered as IPv4, IPv6
 		if primaryProtocol == utiliptables.ProtocolIPv4 {
 			ipt[0] = iptInterface
 			ipt[1] = utiliptables.New(execer, utiliptables.ProtocolIPv6)
 
-			// Just because the feature gate is enabled doesn't mean the node
-			// actually supports dual-stack
-			if _, err := ipt[1].ChainExists(utiliptables.TableNAT, utiliptables.ChainPostrouting); err != nil {
-				klog.Warningf("No iptables support for IPv6: %v", err)
-				dualStack = false
-			}
 		} else {
 			ipt[0] = utiliptables.New(execer, utiliptables.ProtocolIPv4)
 			ipt[1] = iptInterface
 		}
+
+		// check support for ipv6 iptables (used by both ipt and ipvs proxiers).
+		iptIPv6Idx := 1
+		if primaryProtocol == utiliptables.ProtocolIPv4 {
+			iptIPv6Idx = 0
+		}
+		// TODO: @khenidak - should we fail here? Also no checks for single-stack IPv6 clusters.
+		// MUST BE ADDRESSED PART OF KUBE-PROXY DUAL STACK STABLE
+		if _, err := ipt[iptIPv6Idx].ChainExists(utiliptables.TableNAT, utiliptables.ChainPostrouting); err != nil {
```

Comment:
So the reason for the check is because kube-proxy has no idea at this point whether your cluster is actually dual-stack or not, because there is no required kube-proxy argument that would tell it that. (If you had passed a two-valued `--bind-address` or `--cluster-cidr` then it would know from that, but neither of those args is required.)

So kube-proxy just optimistically runs in dual-stack mode, so that it will be ready if it encounters dual-stack services. Except, of course, this totally broke some CI configurations, because they either didn't have `ip6tables` or else had ipv6 disabled in the kernel so `ip6tables` didn't work. (I forget which.)

So now the behavior is "optimistically assume the cluster is dual-stack, except that if `ip6tables` isn't there / doesn't work, then obviously it wasn't supposed to have been dual-stack, so fall back to single-stack ipv4".

We don't check for `ip6tables` in single-stack IPv6 mode, because you can only get single-stack IPv6 by either (a) having no routable IPv4 addresses on the node, or (b) explicitly passing an IPv6 `--bind-address`, so in either case we're not _guessing_ that you want to be single-stack IPv6, we _know_, and so we assume that you configured the node correctly for that.

There should probably be a corresponding check for "does ip[4]tables exist/work", so it can fall back from optimistic-dual-stack to actually-single-stack-ipv6 if necessary as well.

Reply:
I guess since it's still beta, we _could_ change the behavior to "only do dual-stack if the user passed dual `--bind-address`es". That would be a breaking change for existing clusters though so that sounds more like beta2 than GA.

Reply:
It uses node primary address (which is kind of questionable) to identify primary family (this has no baring on anything service specific, just the order of ipt in slice). 

I somehow agree with you. We should test both ipt interfaces. if both available then run with dual stack. if not then print a big warning and run with what we have. It should be only a very dark corner case that shouldn't happen

This is very close to what we have now. But needs a bit of a cleanup.
Kal

Reply:
As most(I think ALL) OSs are dualstack by default, I think kube-proxy should be dual-stack by default (if it can) as it is know.

> It uses node primary address (which is kind of questionable) to identify primary family (this has no baring on anything service specific, just the order of ipt in slice).

kube-proxy doesn't have primary ip family, the nodeIP was needed to get a legit IP when bind-address was 0.0.0.0 
https://github.com/kubernetes/kubernetes/commit/56df70b639c5b5bf53076a3483a2e40b8f60c49a#diff-87f6158dae3a617b9b7941fd224e27e99a56a327902f18b8407b7e859a94c9ad
It seems is required for some cases, I've documented it https://github.com/kubernetes/kubernetes/issues/95515

Reply:
I mean, we don't need `primaryProtocol` anymore, do we?

Reply:
I agree - this all gets simpler if we just try both and see if they work.  This PR or a prequel or a sequel?